### PR TITLE
create, open, sync btc, dcr and ltc wallet features

### PR DESCRIPTION
**Pull request by itswisdomagain**

### CODE REVIEW

**Review by JoeGruffins — APPROVED** [review]
I've only looked at decred, but looks simple and great. So much simpler.

File: asset/dcr/loader.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	if exists, err := WalletExistsAt(params.DataDir); err != nil {
+		return nil, err
+	} else if exists {
+		return nil, fmt.Errorf("wallet at %q already exists", params.DataDir)
+	}
+
+	// Ensure the data directory for the network exists.
+	if err := checkCreateDir(params.DataDir); err != nil {
+		return nil, fmt.Errorf("check new wallet data directory error: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (itswisdomagain) [reply asset/dcr/loader.go:44]: It was borrowed from btc's loader. It's harmless.

File: asset/btc/loader.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	if recovery != nil {
+		seed = recovery.Seed
+	} else {
+		// TODO: Generate seed.
+	}
+
+	// TODO: Encrypt the seed with the private passphrase and save. Should be
+	// able to reveal the seed for this wallet later by providing the private
+	// passphrase. NOTE: cake wallet might require storing the seed without
+	// encrypting first. Insecure, but ...
```
Comment (itswisdomagain) [inline asset/btc/loader.go:53]: A couple TODOs like this relating to seed generation, encryption etc. Resolved in the next PR.

File: asset/btc/loader.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	} else if !exists {
+		return nil, fmt.Errorf("wallet at %q doesn't exist", params.DataDir)
+	}
+
+	chainParams, err := ParseChainParams(params.Net)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing chain params: %w", err)
+	}
+
+	// TODO: Fetch the (encrypted) seed as well.
```
Comment (itswisdomagain) [inline asset/btc/loader.go:113]: Resolved in the next PR.

File: asset/btc/wallet.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	w.log.Debug("Opening BTC wallet...")
+	btcw, err := w.loader.OpenExistingWallet([]byte(wallet.InsecurePubPassphrase), false)
+	if err != nil {
+		return fmt.Errorf("couldn't load wallet: %w", err)
+	}
+
+	neutrinoDBPath := filepath.Join(w.dir, neutrinoDBName)
+	db, err := walletdb.Open(w.dbDriver, neutrinoDBPath, true, dbTimeout) // TODO: DEX uses Create!
```
Comment (itswisdomagain) [inline asset/btc/wallet.go:60]: Need to check if this will cause problems. It shouldn't.

**Review by dreacot — APPROVED** [review]

File: asset/btc/wallet.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+	w.db = db
+	w.mainWallet = btcw
+	return nil
+}
+
+// WalletOpened returns true if the wallet is opened and ready for use.
+func (w *Wallet) WalletOpened() bool {
+	return w.mainWallet != nil
+}
```
[the remark this answers is not in the document]
Comment (itswisdomagain) [reply asset/btc/wallet.go:76]: Yes. The wallet is loaded when other fields are set; and opened when this mainWallet field is set.

File: asset/btc/sync.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	}
+
+	w.log.Debug("Starting neutrino chain service...")
+	chainService, err := neutrino.NewChainService(neutrino.Config{
+		DataDir:       w.dir,
+		Database:      w.db,
+		ChainParams:   *w.chainParams,
+		PersistToDisk: true, // keep cfilter headers on disk for efficient rescanning
+		// AddPeers:      addPeers,
+		// ConnectPeers:  connectPeers,
```
[the remark this answers is not in the document]
Comment (itswisdomagain) [reply asset/btc/sync.go:47]: The spv peer manager borrowed from dex addresses this, I believe.

File: asset/btc/loader.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	chainParams, err := ParseChainParams(params.Net)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing chain: %w", err)
+	}
+
+	if exists, err := WalletExistsAt(params.DataDir); err != nil {
+		return nil, fmt.Errorf("error checking if wallet already exist: %w", err)
+	} else if exists {
+		return nil, fmt.Errorf("wallet at %q already exists", params.DataDir)
+	}
```
[the remark this answers is not in the document]
Comment (itswisdomagain) [reply asset/btc/loader.go:41]: `exists` is only accessible within the if/else statement scope. Replacing the `else if` part with an `if` statement will require defining `exists` outside of the if/else scope. Using `else if` keeps the scope small and I think that's preferable.

**Review by dreacot — APPROVED** [review]

**Pull request merged**